pkg/crypto/ping_slot.go

```diff
@@ -0,0 +1,57 @@
+// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package crypto implements LoRaWAN crypto.
+package crypto
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"encoding/binary"
+	"fmt"
+
+	"go.thethings.network/lorawan-stack/pkg/errors"
+	"go.thethings.network/lorawan-stack/pkg/types"
+)
+
+var pingOffsetCipher cipher.Block
+
+func init() {
+	c, err := aes.NewCipher([]byte{0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00})
+	if err != nil {
+		panic(fmt.Sprintf("failed to create ping offset cipher: %s", err))
+	}
+	pingOffsetCipher = c
+}
+
+const (
+	minPingPeriod = 1 << 5
+	maxPingPeriod = 1 << 12
+)
+
+var errInvalidPingPeriod = errors.DefineInvalidArgument("ping_period", fmt.Sprintf("ping period must be a power of 2 between %d and %d, got {value}", minPingPeriod, maxPingPeriod), "value")
```

Comment:
- no need to make `value` public as it is part of the message format
- backticks for attributes